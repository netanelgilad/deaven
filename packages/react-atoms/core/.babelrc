{
  "plugins": ["babel-plugin-macros"],
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "node": true
        }
      }
    ],
    "@babel/typescript"
  ]
}
