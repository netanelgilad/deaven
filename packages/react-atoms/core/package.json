{
  "name": "@deaven/react-atoms.core",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "author": "Netanel Gilad <netanelgilad@gmail.com>",
  "license": "MIT",
  "scripts": {
    "build": "tsc"
  },
  "dependencies": {
    "react-reconciler": "^0.18.0"
  },
  "devDependencies": {
    "@types/react": "^16.7.20",
    "@types/react-reconciler": "^0.18.0"
  }
}
