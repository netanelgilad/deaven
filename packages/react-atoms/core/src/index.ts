import * as React from "react";

export * from "./render";
export * from "./State";
export * from "./Effect";
export * from "./Ref";
export * from "./NullRenderer";

export function Tuple<T extends React.ReactNode[]>(...props: T) {
  return React.createElement(React.Fragment, null, ...props);
}
