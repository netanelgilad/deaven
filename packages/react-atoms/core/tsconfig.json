{
  "extends": "@deaven/tsconfig/node/tsconfig.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "lib",
    "rootDir": "src",
    "jsx": "react"
  },
  "include": ["src"]
}
