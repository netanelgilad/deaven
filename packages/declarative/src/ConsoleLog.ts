import { useEffect } from "@deaven/react-atoms.core";

type ParmetersOf<T extends (...args: any[]) => any> = T extends (
  ...args: infer U
) => any
  ? U
  : never;

export function ConsoleLog<T>(props: {
  message: T extends any[]
    ? ParmetersOf<typeof console.log>
    : ParmetersOf<typeof console.log>[0];
}) {
  return useEffect(
    () => {
      console.log(props.message);
    },
    [props.message]
  ).render();
}
