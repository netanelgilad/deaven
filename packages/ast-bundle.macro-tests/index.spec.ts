////// <reference path="./external-types" />

import pluginTester from "babel-plugin-tester";
import plugin from "babel-plugin-macros";

function only(strings) {
  return {
    only: true,
    code: strings[0]
  };
}

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    presets: [
      [
        "@babel/preset-env",
        {
          targets: {
            node: true
          }
        }
      ],
      "@babel/typescript"
    ],
    filename: __filename
  },
  tests: [
    `
    import bundle from '@deaven/ast-bundle.macro'

    async function sleep(timeout: number) {
      await new Promise(resolve => setTimeout(resolve, timeout));
    }

    bundle(sleep)
  `,
    `
    import astBundle from "@deaven/ast-bundle.macro";
    import { ConsoleLog } from "./ConsoleLog";
    import { Tuple } from "@deaven/react-atoms.core";

    async function sleep(timeout: number) {
      await new Promise(resolve => setTimeout(resolve, timeout));
    }

    export default function Project() {
      return Tuple(
        ConsoleLog({ message: "Title 1" }),
        ConsoleLog({ message: "Hello World 15" }),
        ConsoleLog({ message: astBundle(sleep) })
      );
    }
    `,
    `
    import bundle from '@deaven/ast-bundle.macro'

    async function sleep(timeout: number) {
      await new Promise(resolve => setTimeout(resolve, timeout));
    }

    bundle(sleep, {export: true})
    `,
    `
    import bundle from '@deaven/ast-bundle.macro'

    const unsafeCast = <T>(obj: any) => {
      return (obj as any) as T;
    }

    bundle(unsafeCast, { export: true })
    `,
    `
    import bundle from '@deaven/ast-bundle.macro'

    export const _ = <T>(arg: T) => arg;
    export const __ = _;

    bundle([_, __], { export: true })
    `
  ]
});
