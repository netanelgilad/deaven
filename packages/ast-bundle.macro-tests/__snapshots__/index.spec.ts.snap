// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros 1. macros: 1. macros 1`] = `
"
import bundle from '@deaven/ast-bundle.macro'

async function sleep(timeout: number) {
  await new Promise(resolve => setTimeout(resolve, timeout));
}

bundle(sleep)

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nasync function sleep(timeout) {\\\\n  await new Promise(resolve => setTimeout(resolve, timeout));\\\\n}\\";
"
`;

exports[`macros 2. macros: 2. macros 1`] = `
"
import astBundle from \\"@deaven/ast-bundle.macro\\";
import { ConsoleLog } from \\"./ConsoleLog\\";
import { Tuple } from \\"@deaven/react-atoms.core\\";

async function sleep(timeout: number) {
  await new Promise(resolve => setTimeout(resolve, timeout));
}

export default function Project() {
  return Tuple(
    ConsoleLog({ message: \\"Title 1\\" }),
    ConsoleLog({ message: \\"Hello World 15\\" }),
    ConsoleLog({ message: astBundle(sleep) })
  );
}

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = Project;

var _ConsoleLog = require(\\"./ConsoleLog\\");

var _reactAtoms = require(\\"@deaven/react-atoms.core\\");

function Project() {
  return (0, _reactAtoms.Tuple)((0, _ConsoleLog.ConsoleLog)({
    message: \\"Title 1\\"
  }), (0, _ConsoleLog.ConsoleLog)({
    message: \\"Hello World 15\\"
  }), (0, _ConsoleLog.ConsoleLog)({
    message: \\"\\\\\\"use strict\\\\\\";\\\\n\\\\nasync function sleep(timeout) {\\\\n  await new Promise(resolve => setTimeout(resolve, timeout));\\\\n}\\"
  }));
}
"
`;

exports[`macros 3. macros: 3. macros 1`] = `
"
import bundle from '@deaven/ast-bundle.macro'

async function sleep(timeout: number) {
  await new Promise(resolve => setTimeout(resolve, timeout));
}

bundle(sleep, {export: true})

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nObject.defineProperty(exports, \\\\\\"__esModule\\\\\\", {\\\\n  value: true\\\\n});\\\\nexports.sleep = sleep;\\\\n\\\\nasync function sleep(timeout) {\\\\n  await new Promise(resolve => setTimeout(resolve, timeout));\\\\n}\\";
"
`;
